sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -157,1069 +55,35 @@ public BlobAsyncClient getSnapshotClient(String snapshot) {
-            values.generateSASQueryParameters(sharedKeyCredential);
-
-        return blobServiceSasQueryParameters.encode();
-    }
-
-    /**
-     * Sets blobServiceSASSignatureValues parameters dependent on the current blob type
-     */
-    BlobServiceSASSignatureValues configureServiceSASSignatureValues(
-        BlobServiceSASSignatureValues blobServiceSASSignatureValues, String accountName) {
-
-        // Set canonical name
-        blobServiceSASSignatureValues.setCanonicalName(this.azureBlobStorage.getUrl(), accountName);
-
-        // Set snapshotId
-        blobServiceSASSignatureValues.setSnapshotId(getSnapshotId());
-
-        // Set resource
-        if (isSnapshot()) {
-            blobServiceSASSignatureValues.setResource(Constants.UrlConstants.SAS_BLOB_SNAPSHOT_CONSTANT);
-        } else {
-            blobServiceSASSignatureValues.setResource(Constants.UrlConstants.SAS_BLOB_CONSTANT);
-        }
-
-        return blobServiceSASSignatureValues;
-    }
-
-    /**
-     * Gets the snapshotId for a blob resource
-     *
-     * @return A string that represents the snapshotId of the snapshot blob
-     */
-    public String getSnapshotId() {
-        return this.snapshot;
-    }
-
-    /**
-     * Determines if a blob is a snapshot
-     *
-     * @return A boolean that indicates if a blob is a snapshot
-     */
-    public boolean isSnapshot() {
-        return this.snapshot != null;
+    public AppendBlobAsyncClient asAppendBlobAsyncClient() {
+        return new SpecializedBlobClientBuilder()
+            .blobAsyncClient(this)
+            .buildAppendBlobAsyncClient();
     }
 
     /**
-     * Get the container name.
+     * Creates a new {@link BlockBlobAsyncClient} associated to this blob.
      *
-     * <p><strong>Code Samples</strong></p>
-     *
-     * {@codesnippet com.azure.storage.blob.BlobAsyncClient.getContainerName}
-     *
-     * @return The name of the container.
+     * @return a {@link BlockBlobAsyncClient} associated to this blob.
      */
-    public final String getContainerName() {
-        return URLParser.parse(this.azureBlobStorage.getUrl(), logger).getContainerName();
+    public BlockBlobAsyncClient asBlockBlobAsyncClient() {
+        return new SpecializedBlobClientBuilder()
+            .blobAsyncClient(this)
+            .buildBlockBlobAsyncClient();
     }
 
     /**
-     * Get the blob name.
+     * Creates a new {@link PageBlobAsyncClient} associated to this blob.
      *
-     * <p><strong>Code Samples</strong></p>
-     *
-     * {@codesnippet com.azure.storage.blob.BlobAsyncClient.getBlobName}
-     *
-     * @return The name of the blob.
+     * @return a {@link PageBlobAsyncClient} associated to this blob.
      */
-    public final String getBlobName() {
-        return URLParser.parse(this.azureBlobStorage.getUrl(), logger).getBlobName();
+    public PageBlobAsyncClient asPageBlobAsyncClient() {
+        return new SpecializedBlobClientBuilder()
+            .blobAsyncClient(this)
```

Comment:
Seeing `this` passed as an argument is iffy to me. You can find yourself having a cyclic dependency graph. I'd prefer to pass in the relevant arguments to create that specialised client (ie. connection string, proxy, transport type, etc. etc.)

Reply:
I agree passing `this` is always terrifying (especially coming from doing a lot of JS development), but the builder API allows passing of a BlobClient and this is the simplest way to go about that.

Reply:
Simplest may not be the best design choice even if the builder supports it because we control the builder API.

Reply:
Changed the builder to allow individual pieces to be passed